core/src/main/java/org/elasticsearch/index/mapper/internal/TypeFieldMapper.java

```diff
@@ -133,12 +134,55 @@ public boolean useTermQueryWithQueryString() {
         @Override
         public Query termQuery(Object value, @Nullable QueryShardContext context) {
             if (indexOptions() == IndexOptions.NONE) {
-                return new ConstantScoreQuery(new PrefixQuery(new Term(UidFieldMapper.NAME, Uid.typePrefixAsBytes(BytesRefs.toBytesRef(value)))));
+                throw new AssertionError();
             }
-            return new ConstantScoreQuery(new TermQuery(createTerm(value)));
+            return new TypeQuery(indexedValueForSearch(value));
         }
     }
 
+    public static class TypeQuery extends Query {
```

Comment:
I see because it is used in tests it has to be public for now...we really need to fix this isolation of classes from there tests, which require so much to be public. But that's unrelated to this change. :)